### PR TITLE
Monitor VmHosts and Vm resources

### LINKED ISSUE #1382: Add monitoring for vm hosts
We already have a monitoring system in place. If the VM host is unreachable for maybe 2 minutes, we can set it to `draining` and create an incident, which helps minimize the impact of the unreachable host incidents.

### PR DESCRIPTION
**Pull request by byucesoy**
This just monitors the VmHost and Vm resources, and creates a page when there is a failure. It does not try to heal it. After using this for a while in prod, we can decide on how the automated healing would look like.

Also fixes #1382

### CODE REVIEW

File: bin/monitor [hunk 1]
```diff
@@ -4,7 +4,7 @@
 require_relative "../loader"
 clover_freeze
 
-MONITORABLE_RESOURCE_TYPES = [PostgresServer, MinioServer, GithubRunner]
+MONITORABLE_RESOURCE_TYPES = [PostgresServer, MinioServer, GithubRunner, VmHost, Vm.left_join(:sshable, id: :id).where(Sequel[:sshable][:id] => nil)]
```
Comment (fdr) [inline bin/monitor:7]: I kind of love this idea of resource types (it's not exactly a type in a conventional sense) being dataset-ish.

**Review by fdr — COMMENTED** [review]
Pretty good, I wonder if `uptime` is what you want considering the value is not used. Using the Good Tool `hyperfine`, there's a cost:

```
fdr@albatross ~/c/clover ((21edc761))> hyperfine -w 3 -N -- true
Benchmark 1: true
  Time (mean ± σ):       1.7 ms ±   0.5 ms    [User: 1.0 ms, System: 0.5 ms]
  Range (min … max):     0.4 ms …   3.0 ms    1763 runs
 
fdr@albatross ~/c/clover ((21edc761)) [1]> hyperfine -w 3 -- cat /proc/sys/kernel/random/boot_id
Benchmark 1: cat
  Time (mean ± σ):       1.9 ms ±   0.8 ms    [User: 1.1 ms, System: 0.9 ms]
  Range (min … max):     0.0 ms …   3.7 ms    702 runs
 
  Warning: Command took less than 5 ms to complete. Note that the results might be inaccurate because hyperfine can not calibrate the shell startup time much more precise than this limit. You can try to use the `-N`/`--shell=none` option to disable the shell completely.
 
fdr@albatross ~/c/clover ((21edc761)) [1]> hyperfine -w 3 -N -- uptime
Benchmark 1: uptime
  Time (mean ± σ):       4.0 ms ±   0.8 ms    [User: 1.2 ms, System: 2.5 ms]
  Range (min … max):     1.1 ms …   5.8 ms    2643 runs
 
```

If you really, really don't care about output, `true` does the trick.

I included `cat` of the boot-id because we use that elsewhere to monitor whether the system is in the same boot as we expected/previously. It's almost as fast as "true," and it has the disadvantage of shell execution here. Not sure if that gives you any ideas for this part of the program or consolidation opportunities to reduce code, but I threw it in for comparison sake.

Possibly the shell expression `true` is even faster than what hyperfine (which is designed to put programs under test), because shells will often avoid forking the whole `true` program for it (which is order of magnitudes slower:

```
fdr@albatross ~/c/clover ((21edc761))> time uptime
 10:31:51 up 4 days, 13 min,  2 users,  load average: 1.45, 1.90, 2.02

________________________________________________________
Executed in    4.97 millis    fish           external
   usr time    1.03 millis    1.03 millis    0.00 millis
   sys time    4.00 millis    0.29 millis    3.70 millis

fdr@albatross ~/c/clover ((21edc761))> time true

________________________________________________________
Executed in    6.00 micros    fish           external
   usr time    6.00 micros    6.00 micros    0.00 micros
   sys time    1.00 micros    1.00 micros    0.00 micros

```

File: prog/vm/host_nexus.rb [hunk 2]
```diff
@@ -206,11 +206,25 @@ def before_run
...
+      hop_unavailable if !available?
+    end
+
     Clog.emit("vm host utilization") { {vm_host_utilization: vm_host.values.slice(:location, :arch, :total_cores, :used_cores, :total_hugepages_1g, :used_hugepages_1g, :total_storage_gib, :available_storage_gib).merge({vms_count: vm_host.vms_dataset.count})} }
 
     nap 30
   end
 
+  label def unavailable
+    Prog::PageNexus.assemble("#{vm_host} is unavailable", vm_host.ubid, "VmHostUnavailable", vm_host.ubid)
```
Comment (enescakir) [inline prog/vm/host_nexus.rb:219]: I designed it poorly. `related_resources` should be an array.

Comment (byucesoy) [reply prog/vm/host_nexus.rb:219]: I made a small chance to PageNexus.assemble. How does it look like now?

**Comment by byucesoy** [comment]
> If you really, really don't care about output, `true` does the trick.

I really don't care about the output. At least for now. Maybe in the future, we can use this to collect some information, but for now `true` does the trick, I changed it.

File: prog/page_nexus.rb [hunk 3]
```diff
@@ -5,6 +5,7 @@ class Prog::PageNexus < Prog::Base
   semaphore :resolve
 
   def self.assemble(summary, related_resources, *tag_parts)
+    related_resources = [related_resources] unless related_resources.is_a?(Array)
```
Comment (enescakir) [inline prog/page_nexus.rb:8]: `Array(related_resources)` behaves exact same.

File: prog/vm/nexus.rb [hunk 4]
```diff
@@ -490,6 +490,14 @@ def before_run
       hop_update_firewall_rules
     end
 
+    when_checkup_set? do
+      hop_unavailable if !available?
+    rescue Sshable::SshError
+      # Host is down, which will be handled by HostNexus
+      # No need to create a page for each Vm separately.
+      decr_checkup
```
Comment (enescakir) [inline prog/vm/nexus.rb:498]: Should we hop to the `unavailable` label and decide not to create a page there? This way, both the operator and the customer can see that the VM is unavailable.

File: prog/vm/host_nexus.rb [hunk 5]
```diff
@@ -206,11 +206,25 @@ def before_run
...
+    when_checkup_set? do
+      hop_unavailable if !available?
+    end
+
     Clog.emit("vm host utilization") { {vm_host_utilization: vm_host.values.slice(:location, :arch, :total_cores, :used_cores, :total_hugepages_1g, :used_hugepages_1g, :total_storage_gib, :available_storage_gib).merge({vms_count: vm_host.vms_dataset.count})} }
 
     nap 30
   end
 
+  label def unavailable
```
Comment (enescakir) [inline prog/vm/host_nexus.rb:218]: Should we automatically update the allocation_state to "draining", or should the on-call engineer handle it?

**Review by enescakir — COMMENTED** [review]
Could you provide some explanation in the commit message as to why we only select virtual machines that have no sshable?
I assume because the virtual machines with sshable are used for our other services and higher level services already monitored

**Comment by fdr** [comment]
> Could you provide some explanation in the commit message as to why we only select virtual machines that have no sshable? I assume because the virtual machines with sshable are used for our other services and higher level services already monitored

That's a good question...and plausible guess, but the reader should not have to guess if they were to read the blame I think, or possibly even a comment, this seems like a stable enough fact related to the expression that's unlikely to bitrot.

Now that I think about it, it's not always easy to tell how the optimizer will regard this left join (it actually depends on NOT NULL constraints), to make sure it interprets it as the (efficient) antijoin, the more "direct" route (not syntactically, but perhaps logically, avoiding [vagaries of SQL NULL](http://thoughts.davisjeff.com/2009/08/02/what-is-the-deal-with-nulls/)) is a correlated subquery using the `EXISTS`. https://stackoverflow.com/questions/68407054/fastest-way-to-anti-join-in-postgres.

Perplexity sez it looks a bit like this:

```ruby
User.where(Sequel.negate(Sequel.expr(Sequel.exists(Order.where(Order[:user_id] => Sequel[:users][:id])))))
```

File: prog/vm/host_nexus.rb (hunk 5 above)
Comment (byucesoy) [reply prog/vm/host_nexus.rb:218]: I prefer to leave this to on-call for a while, only to see failure modes and gain some understanding. I don't have any objection for updating the allocation_state (or some other ways of automated action) after that.

**Comment by byucesoy** [comment]
> Could you provide some explanation in the commit message as to why we only select virtual machines that have no sshable? I assume because the virtual machines with sshable are used for our other services and higher level services already monitored

> That's a good question...and plausible guess, but the reader should not have to guess if they were to read the blame...or a comment. (somehow this feels unlikely to decay into misinformation quickly, but you tell me, in such cases I consider a comment)

Yep, the guess is correct. I added below section to the commit message.

```
For monitoring purposes, we only focus on Vms that does not have an associated
Sshable entity. This is because Vms with Sshable entity are childs of higher
level resources like PostgresServer or GithubRunner and will be monitored by
the their respective parent resource. This is a safeguard against duplicate
page creation for the same problem. When we are much bigger organization where
there are specialized teams for each resource, it would make sense to create
separate pages for Vms and their parent resources, because each team would
independently try to heal their resources. However, for now, this is not
necessary.
```

**Comment by byucesoy** [comment]
> Now that I think about it, it's not always easy to tell how the optimizer will regard this left join (it actually depends on NOT NULL constraints), to make sure it interprets it as the (efficient) antijoin, the more "direct" route (not syntactically, but perhaps logically, avoiding [vagaries of SQL NULL](http://thoughts.davisjeff.com/2009/08/02/what-is-the-deal-with-nulls/)) is a correlated subquery using the EXISTS. https://stackoverflow.com/questions/68407054/fastest-way-to-anti-join-in-postgres.
> Perplexity sez it looks a bit like this:
> User.where(Sequel.negate(Sequel.expr(Sequel.exists(Order.where(Order[:user_id] => Sequel[:users][:id])))))

I simplified it to this;

`Vm.where(~Sshable.where(id: Sequel[:vm][:id]).exists)`

File: prog/vm/nexus.rb (hunk 4 above)
Comment (byucesoy) [reply prog/vm/nexus.rb:498]: makes sense, changed it.

**Pull request merged**